ui/src/main/java/edu/wpi/grip/ui/DeployController.java

```diff
@@ -0,0 +1,216 @@
+        address.setText(settings.getDeployAddress());
+        user.setText(settings.getDeployUser());
+        javaHome.setText(settings.getDeployJavaHome());
+        deployDir.setText(settings.getDeployDir());
+    }
+
+    private void saveSettings() {
+        // If the settings are updated in the deploy dialog, we still want to save them in the persistent project
+        // settings, so they don't get reset the next time the project is opened to the settings are edited.
+        final ProjectSettings settings = pipeline.getProjectSettings();
+        settings.setDeployAddress(address.getText());
+        settings.setDeployUser(user.getText());
+        settings.setDeployJavaHome(javaHome.getText());
+        settings.setDeployDir(deployDir.getText());
+        eventBus.post(new ProjectSettingsChangedEvent(settings));
+    }
+
+    @FXML
+    public void onDeploy() {
+        saveSettings();
+
+        deploying.setValue(true);
+        console.clear();
+
+        // Start the deploy in a new thread, so the GUI doesn't freeze
+        deployThread = Optional.of(new Thread(() ->
+                deploy(address.getText(), user.getText(), password.getText(), javaHome.getText(), deployDir.getText())));
+        deployThread.get().setDaemon(true);
+        deployThread.get().start();
+    }
+
+    @FXML
+    public void onStop() {
+        deployThread.ifPresent(Thread::interrupt);
+        deploying.setValue(false);
+        status.setText("");
+    }
+
+    /**
+     * Upload and run the GRIP project using the current deploy settings.  This is run in a separate thread, and it
+     * periodically updates the GUI to inform the user of the current status of the deployment.
+     */
+    private void deploy(String address, String user, String password, String javaHome, String deployDir) {
+        setStatusAsync("Connecting to " + address, false);
+
+        try (SSHClient ssh = new SSHClient()) {
+            ssh.loadKnownHosts();
+            ssh.connect(address);
+            ssh.authPassword(user, password);
+
+            // Update the progress bar and status text while uploading files
+            SCPFileTransfer scp = ssh.newSCPFileTransfer();
+            scp.setTransferListener(new LoggingTransferListener() {
+                @Override
+                public StreamCopier.Listener file(String name, long size) {
+                    setStatusAsync("Uploading " + name, false);
+                    return transferred -> {
+                        if (isNotCanceled())
+                            Platform.runLater(() -> progress.setProgress((double) transferred / size));
+                    };
+                }
+            });
+
+            // The project may or may not be saved to a file (and even if it was saved, it might be modified), so we
+            // serialize it to a string before deploying.
+            StringWriter projectWriter = new StringWriter();
+            project.save(projectWriter);
```

Comment:
Sure. Thats fine. But shouldn't you offer them the option of saving before deploying.